php/libraries/Database.class.inc

```diff
@@ -1159,5 +1159,83 @@ class Database
         }
     }
 
+    /**
```

Comment:
I think these changes should be a different PR. I have nothing against adding transaction support, but it's a completely different concern.

(This code mostly looks fine. I'm not thrilled about the fact that it heavily depends upon exceptions for control flow even in non-exceptional circumstances, but I haven't looked into if PDO has a better way to do it.)